homeassistant/components/light/pilight.py

```diff
@@ -0,0 +1,207 @@
+}, extra=vol.ALLOW_EXTRA)
+
+RECEIVE_SCHEMA = COMMAND_SCHEMA.extend({
+    vol.Optional(CONF_ECHO): cv.boolean
+})
+
+SWITCHES_SCHEMA = vol.Schema({
+    vol.Required(CONF_ON_CODE): COMMAND_SCHEMA,
+    vol.Required(CONF_OFF_CODE): COMMAND_SCHEMA,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_OFF_CODE_RECIEVE, default=[]): vol.All(cv.ensure_list,
+                                                             [COMMAND_SCHEMA]),
+    vol.Optional(CONF_ON_CODE_RECIEVE, default=[]): vol.All(cv.ensure_list,
+                                                            [COMMAND_SCHEMA]),
+    vol.Optional(CONF_DIMMABLE, default=False): cv.boolean
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_DEVICES):
+        vol.Schema({cv.string: SWITCHES_SCHEMA}),
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Pilight platform."""
+    devices = []
+
+    for dev_name, properties in config.get(CONF_DEVICES).items():
+        devices.append(
+            PilightDimmableLight(
+                hass,
+                properties.get(CONF_NAME, dev_name),
+                properties.get(CONF_ON_CODE),
+                properties.get(CONF_OFF_CODE),
+                properties.get(CONF_ON_CODE_RECIEVE),
+                properties.get(CONF_OFF_CODE_RECIEVE),
+                properties.get(CONF_DIMMABLE)
+            )
+        )
+
+    add_devices(devices)
+
+
+class _ReceiveHandle(object):
+    def __init__(self, config, echo):
+        """Initialize the handle."""
+        self.config_items = config.items()
+        self.echo = echo
+
+    def match(self, code):
+        """Test if the received code matches the configured values.
+
+        The received values have to be a subset of the configured options.
+        """
+        return self.config_items <= code.items()
+
+    def run(self, light, turn_on):
+        """Change the state of the switch."""
+        light.set_state(turn_on=turn_on, send_code=self.echo)
+
+
+class PilightDimmableLight(Light):
+    """Representation of a Pilight light."""
+
+    def __init__(self, hass, name, code_on, code_off, code_on_receive,
+                 code_off_receive, dimmable):
+        """Initialize the light."""
+        self._hass = hass
+        self._name = name
+        self._state = False
+
+        self._code_on = code_on
+        self._code_off = code_off
+
+        self._code_on_receive = []
+        self._code_off_receive = []
+
+        self._brightness = 255
+        self._dimmable = dimmable
+
+        for code_list, conf in ((self._code_on_receive, code_on_receive),
+                                (self._code_off_receive, code_off_receive)):
+            for code in conf:
+                echo = code.pop(CONF_ECHO, True)
+                code_list.append(_ReceiveHandle(code, echo))
+
+        if any(self._code_on_receive) or any(self._code_off_receive):
+            hass.bus.listen(pilight.EVENT, self._handle_code)
```

Comment:
You can not use the Event bus. We remove that feature in future or a add a whitelist for it.

Reply:
He copied this from https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/pilight.py#L123 as the pilight-events get distributed as events.

See https://github.com/home-assistant/home-assistant/pull/7344 and https://github.com/home-assistant/home-assistant/issues/7372#issuecomment-298264660

Reply:
We should not add new code that uses the eventbus for device communication. So refactoring pilight to use `dispatcher` and the fire certain codes as events code is a prerequisite.